compiler/qsc_linter/src/lints/hir.rs

```diff
@@ -1,17 +1,104 @@
-use qsc_hir::hir::Lit;
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use qsc_hir::{
+    hir::{CallableDecl, CallableKind, Expr, ExprKind, SpecBody, SpecDecl, Stmt, StmtKind},
+    ty::Ty,
+    visit::{self, Visitor},
+};
 
 use crate::linter::hir::declare_hir_lints;
 
 use super::lint;
 
 declare_hir_lints! {
-    (Placeholder, LintLevel::Allow, "this a placeholder", "remove after addding the first HIR lint"),
+    (NeedlessOperation, LintLevel::Warn, "unnecessary operation declaration", "convert to function")
+}
+
+// Helper to check if a operation has desired operation characteristics
+//
+// empty operations: no lint, special case of `I` operation used for Delay
+// operations with errors (e.g. partially typed code): no lint because linter does not run
+// non-empty operations, with specializations, and no quantum operations: show lint, but don't offer quickfix (to avoid deleting user code in any explicit specializations)
+// non-empty operations with no specializations, and no quantum operations: show lint, offer quickfix to convert to function
+#[derive(Default)]
+struct OperationLimits {
+    // Operation Characteristics
+    op_char: bool,
 }
```

Comment:
Maybe we could name this struct something like `IsQuantumOperation` to convey its meaning better (or something along those lines). I didn't really get what "limits" or "op_char" meant when reading the code the first time.

```suggestion
struct OperationLimits {
    // Operation Characteristics
    op_char: bool,
}
```
```suggestion
struct IsQuantumOperation(bool);
```

Reply:
Yes, that sounds good!